content/blog/2020-11-09-modern-gc/index.md

```diff
@@ -0,0 +1,80 @@
+
+Delay reference counting can then be implemented in the conservative Immix collector by treating each ambigious reference as a possible root. Instead of incrementing a counter on each allocation, you wait to do the counting of allocations pointed to from a possible root until you need to garbage collect. The method is then as follows, get the list of ambigious references to valid objects, increment the lines for each root reference it has, free all zero count lines. Before the next round of garbage collection, decrement all of the lines that were increment this round by the amount they were incremented. In a way, this is like tracing only at the topmost objects of the heap. Then, only if one of the topmost objects dies to you iterate through it's tree decrementing the reference count of lines looking for more objects to free.
+
+## Evaluation
+
+The evaluation for this paper is rather extensive and follows many of the guidelines we would hope for such as:
+
+- describing the choice of benchmarks and why some were left out
+- specifying the java virtual machine used
+- parameters like block and line size
+- warmup periods
+- a description of the hardware and software platforms being used
+- reporting both averages and confidence intervals
+
+The goal of their analysis is to show that conservative garbage collectors are not as limited as they might first appear and that a conservative garbage collector can be competitive against the state of the art jvm garbage collector.
+
+### Impact of Conservatism
+
+Probably the most obvious concern with conservative garbage collection is that dead objects will survive in the heap due to an ambiguous reference that is actually a value. However in practice, the paper found that these zombie references are not much of a concern. They measured the size of memory retained by undead objects over the memory size of live objects and found the average to be 0.02%. This would suggest that in most cases, the amount of extra objects maintained due to an ambiguous reference is miniscule.
+
+Similarly, since every ambiguous reference must be filtered against the object map, there may be a concern with the amount of overhead added to be conservative. Comparing exact Mark-Sweep with conservative Mark-Sweep, the paper found a 19.8% increase in overhead when garbage collecting. However, the effect on total runtime of the program was 3.7% which suggests that the overhead for just having to filter ambiguous pointers isn't huge.
```

Comment:
run time, or running time https://capra.cs.cornell.edu/styleguide/#runtime

Reply:
Sounds like there should be a Capra style guide linter